### PR TITLE
Add velero deployment controller

### PR DESCRIPTION
**Pull request by shubham-pampattiwar**
This PR adds the Velero deployment controller

### CODE REVIEW

File: oadp-operator/controllers/velero.go [hunk 1]
```diff
@@ -1,14 +1,469 @@
...
 	}
+
+	// Build Velero Deployment
+	veleroDeployment := r.buildVeleroDeployment(&velero)
+
+	// Create or Update Velero Deployment
+	op, err := controllerutil.CreateOrUpdate(r.Context, r.Client, &veleroDeployment, func() error {
+
+		// Set controller reference to Velero controller
+		err := controllerutil.SetControllerReference(&velero, &veleroDeployment, r.Scheme)
```
Comment (shawn-hurley) [inline oadp-operator/controllers/velero.go:72]: I don't think that this is changing anything in the cluster just on the object reference

File: oadp-operator/controllers/velero.go [hunk 2]
```diff
@@ -1,14 +1,469 @@
...
+			corev1.EventTypeNormal,
+			"VeleroDeploymentReconciled",
+			fmt.Sprintf("performed %s on velero deployment %s/%s", op, veleroDeployment.Namespace, veleroDeployment.Name),
+		)
+	}
 	return true, nil
 }
+
+// Build Velero Deployment
+func (r *VeleroReconciler) buildVeleroDeployment(velero *oadpv1alpha1.Velero) appsv1.Deployment {
```
Comment (alaypatel07) [inline oadp-operator/controllers/velero.go:99]: return a pointer? `*appsv1.Deployment`

Comment (dymurray) [reply oadp-operator/controllers/velero.go:99]: +1

File: oadp-operator/controllers/velero.go [hunk 3]
```diff
@@ -1,14 +1,469 @@
...
+	veleroResourceReqs := r.getVeleroResourceReqs(velero)
+
+	deployment := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      Velero,
+			Namespace: VeleroNamespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			//TODO: add velero nodeselector, needs to be added to the Velero CR first
+			Selector: &metav1.LabelSelector{
```
Comment (alaypatel07) [inline oadp-operator/controllers/velero.go:114]: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates

I think you will have to be careful here. You might want to accept the pointer to deployment here and change the parts that you need based on whether it's a create or update.

File: oadp-operator/controllers/velero.go [hunk 4]
```diff
@@ -1,14 +1,469 @@
...
+	VeleroNamespace = "oadp-operator"
+	//TODO: Check for default secret names
+	VELERO_AWS_SECRET_NAME   = "cloud-credentials"
+	VELERO_AZURE_SECRET_NAME = "cloud-credentials-azure"
+	VELERO_GCP_SECRET_NAME   = "cloud-credentials-gcp"
+)
+
+// Environment Vars
+const (
+	LD_LIBRARY_PATH               = "LD_LIBRARY_PATH"
```
Comment (dymurray) [inline oadp-operator/controllers/velero.go:27]: +1 good to make these constants

File: oadp-operator/controllers/velero.go [hunk 5]
```diff
@@ -1,14 +1,469 @@
...
+	GCP_SHARED_CREDENTIALS_FILE   = "GCP_SHARED_CREDENTIALS_FILE"
+	//TODO: Add Proxy env vars
+	HTTP_PROXY  = "HTTP_PROXY"
+	HTTPS_PROXY = "HTTPS_PROXY"
+	NO_PROXY    = "NO_PROXY"
+)
+
+//TODO: Add Image customization options
+// Images
+const (
```
Comment (dymurray) [inline oadp-operator/controllers/velero.go:41]: NIT: you will need to make the tags and images constants we can override with env vars

Comment (dymurray) [reply oadp-operator/controllers/velero.go:41]: We can do this in a future update though

File: oadp-operator/controllers/velero.go (hunk 3 above)
Comment (dymurray) [reply oadp-operator/controllers/velero.go:114]: +1 you could be deleting labels by doing this

File: oadp-operator/controllers/velero.go [hunk 6]
```diff
@@ -1,14 +1,464 @@
...
 		return false, err
 	}
+
+	veleroVolumeMounts := r.getVeleroVolumeMounts(&velero)
+	veleroVolumes := r.getVeleroVolumes(&velero)
+	veleroEnv := r.getVeleroEnv(&velero)
+	veleroInitContainers := r.getVeleroInitContainers(&velero)
+	veleroResourceReqs := r.getVeleroResourceReqs(&velero)
+	veleroTolerations := velero.Spec.VeleroTolerations
+	veleroDeployment := &appsv1.Deployment{}
```
Comment (shawn-hurley) [inline oadp-operator/controllers/velero.go:71]: I think this may need the namespace and name

File: oadp-operator/controllers/velero.go [hunk 7]
```diff
@@ -1,14 +1,486 @@
...
+	//TODO: Check for default secret names
+	VELERO_AWS_SECRET_NAME   = "cloud-credentials"
+	VELERO_AZURE_SECRET_NAME = "cloud-credentials-azure"
+	VELERO_GCP_SECRET_NAME   = "cloud-credentials-gcp"
+)
+
+// Environment Vars
+const (
+	LD_LIBRARY_PATH               = "LD_LIBRARY_PATH"
+	VELERO_NAMESPACE              = "VELERO_NAMESPACE"
```
Comment (shawn-hurley) [inline oadp-operator/controllers/velero.go:30]: shouldn't this be "velero"?

File: oadp-operator/controllers/velero.go [hunk 8]
```diff
@@ -1,14 +1,486 @@
...
+		}
+
+		// Setting controller owner reference on the velero deployment
+		err := controllerutil.SetControllerReference(&velero, veleroDeployment, r.Scheme)
+		if err != nil {
+			return err
+		}
+
+		// update the Deployment template
+		veleroDeployment = r.buildVeleroDeployment(veleroDeployment, veleroLabels, veleroVolumeMounts, veleroVolumes, veleroEnv, veleroInitContainers, veleroResourceReqs, veleroTolerations)
```
Comment (alaypatel07) [inline oadp-operator/controllers/velero.go:99]: an opinionated suggestion: this function call is not providing a lot of values, you are sending the arg list into the function and it is just setting it to the right place in deployment. 

Why don't we change this to `r.buildVeleroDeployment(veleroDeployment)`? then the function will call all the getters and build the right spec there. Then we can add a unit test for `buildVeleroDeployment`. 

Not asking for the units in this PR but let's start with refactoring this function so it more makes

File: oadp-operator/controllers/velero.go (hunk 7 above)
Comment (dymurray) [reply oadp-operator/controllers/velero.go:30]: This is the env var key used on the container. The var name may want to communicate that

File: oadp-operator/controllers/velero.go (hunk 8 above)
Comment (dymurray) [reply oadp-operator/controllers/velero.go:99]: Agree with this. Generally my rule of thumb is anything over 3 args means we can refactor the function call in some manner. I think Alay's suggestion makes sense

**Review by dymurray — APPROVED** [review]
Other than the NITS I left, this LGTM

**Review by alaypatel07 — APPROVED** [review]
Looks great.  Thanks!

**Review by dymurray — APPROVED** [review]

**Pull request merged**